src/py/flwr/server/driver/inmemory_driver.py

```diff
@@ -43,37 +44,47 @@ class InMemoryDriver(Driver):
 
     def __init__(
         self,
+        run_id: int,
         state_factory: StateFactory,
-        fab_id: Optional[str] = None,
-        fab_version: Optional[str] = None,
     ) -> None:
-        self.run_id: Optional[int] = None
-        self.fab_id = fab_id if fab_id is not None else ""
-        self.fab_version = fab_version if fab_version is not None else ""
+        self._run_id = run_id
+        self._fab_id = ""
+        self._fab_ver = ""
+        self._has_initialized = False
         self.node = Node(node_id=0, anonymous=True)
         self.state = state_factory.state()
 
     def _check_message(self, message: Message) -> None:
         # Check if the message is valid
         if not (
-            message.metadata.run_id == self.run_id
+            message.metadata.run_id == self.run.run_id
             and message.metadata.src_node_id == self.node.node_id
             and message.metadata.message_id == ""
             and message.metadata.reply_to_message == ""
             and message.metadata.ttl > 0
         ):
             raise ValueError(f"Invalid message: {message}")
 
-    def _get_run_id(self) -> int:
-        """Return run_id.
-
-        If unset, create a new run.
-        """
-        if self.run_id is None:
-            self.run_id = self.state.create_run(
-                fab_id=self.fab_id, fab_version=self.fab_version
-            )
-        return self.run_id
+    def _init_run(self) -> None:
+        """Initialize the run."""
+        if self._has_initialized:
+            return
+        run = self.state.get_run(self._run_id)
+        if run is None:
+            raise RuntimeError(f"Cannot find the run with ID: {self._run_id}")
+        self._fab_id = run.fab_id
+        self._fab_ver = run.fab_version
+        self._has_initialized = True
+
+    @property
+    def run(self) -> Run:
+        """Run ID."""
+        self._init_run()
+        return Run(
+            run_id=self._run_id,
+            fab_id=self._fab_id,
+            fab_version=self._fab_ver,
+        )
```

Comment:
a higher level question: why not creating the `Run` obj in `__init__()`? will help simplify this logic a bit imo

Reply:
I just find having to go through `_init_run()` (even if returning immediately) each time this class access `self.run` a bit unnecessary. Or is this done in this way to support some upcoming functionality?

Reply:
I think it's a rather common property implementation (though maybe not in Python?). We also had the same workflow before this PR, e.g., `run_id = self._get_run_id()`. Regarding creating a `Run` object in the `run` property each time, it's to prevent the `Run` object from being modified externally.

As for creating the `Run` object in `__init__()`, I think we should postpone the initialization until necessary, following the same workflow we had previously.

Reply:
yeah, we have the same logic before this PR. I thought it would make it unnecessarily complex. But i see the value of _"prevent the Run object from being modified externally."_. I'll resolve this comment.